node/src/asset_cleanup.rs

```diff
@@ -0,0 +1,416 @@
+    let Some(db_res): Option<Vec<u8>> = db.get(DBCol::EpochData, EPOCH_ID_KEY)? else {
+        return Ok(None);
+    };
+    if let Ok(epoch_data) = bincode::deserialize::<EpochData>(&db_res) {
+        return Ok(Some(EpochDataWrapper::Current(epoch_data)));
+    }
+
+    if db_res.len() == 8 {
+        let bytes_array: [u8; 8] = db_res
+            .as_slice()
+            .try_into()
+            .inspect_err(|bytes| tracing::error!("PREVIOUS EPOCH_ID ENTRY NOT u64: {:?}", bytes))?;
+        let epoch_id_number = u64::from_be_bytes(bytes_array);
+        let epoch_id = EpochId::new(epoch_id_number);
+        return Ok(Some(EpochDataWrapper::Legacy(epoch_id)));
+    };
+    anyhow::bail!("Can't deserialize EPOCH_ID entry: {:?}", db_res);
+}
+
+enum AssetCleanup {
+    Keep,
+    DeleteAll,
+    KeepOnly(Vec<primitives::ParticipantId>),
+}
+
+fn get_cleanup_behavior(
+    previous_epoch_data: &EpochData,
+    current_epoch_data: &EpochData,
+) -> AssetCleanup {
+    if previous_epoch_data.epoch_id != current_epoch_data.epoch_id {
+        tracing::info!(
+            "Deleting presignatures and triples. NEW EPOCH: {:?}, OLD EPOCH {:?}.",
+            current_epoch_data.epoch_id,
+            previous_epoch_data.epoch_id
+        );
+        return AssetCleanup::DeleteAll;
+    }
+    tracing::info!("Same Epoch Id.");
+
+    let persitent_participants: Vec<primitives::ParticipantId> = current_epoch_data
+        .participants
+        .participants
+        .iter()
+        .filter_map(
+            |current| match previous_epoch_data.participants.get_info(current.id) {
+                None => Some(current.id),
+                Some(prev) => (current.p2p_public_key == prev.p2p_public_key).then_some(prev.id),
+            },
+        )
+        .collect();
+
+    let tls_key_changed: bool =
+        persitent_participants.len() != current_epoch_data.participants.participants.len();
+    if tls_key_changed {
+        AssetCleanup::KeepOnly(persitent_participants)
+    } else {
+        AssetCleanup::Keep
+    }
+}
+
+fn get_cleanup_behavior_during_update(
+    previous_epoch_id: EpochId,
+    current_epoch_id: EpochId,
+) -> AssetCleanup {
+    tracing::info!("Updating from legacy node");
+    if previous_epoch_id != current_epoch_id {
+        tracing::info!(
+            "We should not be updating during a resharing, but we will try our best. Current epoch id: {:?}, old epoch id {:?}.",
+            current_epoch_id, previous_epoch_id
+        );
+        AssetCleanup::DeleteAll
+    } else {
+        AssetCleanup::Keep
+    }
+}
+
+// This function compares the `current_epoch_data` with the epoch data stored in the database.
+// If the epoch id changed, then all assets (triples & presignatures) are deleted.
+// If the epoch id did not change, then the node will discard all owned assets (triples &
+// presignatures) involving any nodes that changed their TLS key.
+pub fn delete_stale_triples_and_presignatures(
```

Comment:
Code organization nit: This feels like the most significant function in this module so I'd find it more readable to have it defined first.